storage.go

```diff
@@ -3,34 +3,187 @@ package ftp
+	counter           int
+	buf               []byte
+}
+
 func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
-	panic("not implemented")
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o = s.newObject(false)
+		o.Mode = ModeDir
+	} else {
+		o = s.newObject(false)
+		o.Mode = ModeRead
+	}
+
+	o.ID = filepath.Join(s.workDir, path)
+	o.Path = path
+	return o
+}
+func (s *Storage) createDir(ctx context.Context, path string) (o *Object, err error) {
+	rp := s.getAbsPath(path)
+	s.connection.ChangeDir(s.workDir)
+	err = s.connection.MakeDir(rp)
+	if err != nil {
+		return
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+	o.Mode |= ModeDir
+	return
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	s.connection.ChangeDir(filepath.Dir(rp))
+	err = s.connection.Delete(filepath.Base(rp))
+	s.connection.ChangeDir(s.workDir)
+	if err != nil {
+		return err
+	}
+	return
+}
+func (input *listDirInput) ContinuationToken() string {
+	return input.continuationToken
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	//rp := s.getAbsPath(path)
+	input := listDirInput{
+		// Always keep service original name as rp.
+		rp: s.getAbsPath(path),
+		// Then convert the dir to slash separator.
+		dir: filepath.ToSlash(path),
+
+		// if HasContinuationToken, we should start after we scanned this token.
+		// else, we can start directly.
+		started:           !opt.HasContinuationToken,
+		continuationToken: opt.ContinuationToken,
+
+		buf: make([]byte, 8192),
+	}
+
+	return NewObjectIterator(ctx, s.listDirNext, &input), nil
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.WorkDir = s.workDir
+	return meta
+
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if err != nil {
+		return 0, err
+	}
+
+	r, err := s.connection.Retr(rp)
+	if opt.HasOffset {
+		r, err = s.connection.RetrFrom(rp, uint64(opt.Offset))
+		if err != nil {
+			return n, err
+		}
+	}
+	defer func() {
+		closeErr := r.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	// if opt.HasIoCallback {
+	// 	r = iowrap.CallbackReadCloser(*r, opt.IoCallback)
+	// }
+	if opt.HasSize {
+		return io.CopyN(w, r, opt.Size)
+	}
+	return io.Copy(w, r)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	fl, err := s.connection.List(filepath.Dir(rp))
+	if err != nil {
+		return nil, err
+	}
+	var fe *ftp.Entry = nil
+	for i := range fl {
+		if fl[i].Name == filepath.Base(rp) {
+			fe = fl[i]
+		}
+	}
+	if fe == nil {
+		return nil, err
```

Comment:
It seems like this err will always be nil?

Reply:
Oh I see, maybe it is more suitable to return a `ErrObjectNotExist` error here, check [errors.go](https://github.com/beyondstorage/go-storage/blob/master/services/error.go#L46) as reference.